File: samples/language-impl-comparison/elixir/mini_lisp.exs
defmodule MiniLisp do
  @moduledoc """
  ミニ Lisp 評価機。S 式をトークン列に分割し、再帰下降で解析・評価する。
  """

  @type expr ::
          {:number, float()}
          | {:symbol, String.t()}
          | {:list, [expr()]}

  @type value ::
          {:number, float()}
          | {:lambda, [String.t()], expr(), env()}
          | {:builtin, ( [value()] -> result(value()) )}

  @type env :: %{optional(String.t()) => value()}
  @type result(t) :: {:ok, t} | {:error, String.t()}

  @spec eval(String.t()) :: result(value())
  def eval(source) do
    tokens = tokenize(source)

    with {:ok, {expr, []}} <- parse_expr(tokens),
         {:ok, value} <- eval_expr(expr, default_env()) do
      {:ok, value}
    else
      {:ok, {_expr, rest}} -> {:error, "未消費トークンがあります: #{Enum.join(rest, " ")}"}
      {:error, reason} -> {:error, reason}
    end
  end

  @spec tokenize(String.t()) :: [String.t()]
  def tokenize(source) do
    source
    |> String.replace("(", " ( ")
    |> String.replace(")", " ) ")
    |> String.split()
  end

  @spec parse_expr([String.t()]) :: result({expr(), [String.t()]})
  def parse_expr([]), do: {:error, "入力が空です"}

  def parse_expr(["(" | rest]), do: parse_list(rest, [])

  def parse_expr([")" | _]), do: {:error, "対応しない閉じ括弧です"}

  def parse_expr([token | rest]) do
    {:ok, {atom(token), rest}}
  end

  defp atom(token) do
    case Float.parse(token) do
      {number, ""} -> {:number, number}
      _ -> {:symbol, token}
    end
  end

  defp parse_list([], _acc), do: {:error, "リストが閉じられていません"}

  defp parse_list([")" | rest], acc), do: {:ok, {{:list, Enum.reverse(acc)}, rest}}

  defp parse_list(tokens, acc) do
    with {:ok, {expr, rest}} <- parse_expr(tokens) do
      parse_list(rest, [expr | acc])
    end
  end

  defp eval_expr({:number, n}, _env), do: {:ok, {:number, n}}

  defp eval_expr({:symbol, name}, env) do
    case Map.fetch(env, name) do
      {:ok, value} -> {:ok, value}
      :error -> {:error, "未定義シンボル: #{name}"}
    end
  end

  defp eval_expr({:list, []}, _env), do: {:error, "空の式は評価できません"}

  defp eval_expr({:list, [{:symbol, "lambda"}, {:list, params}, body]}, env) do
    with {:ok, names} <- ensure_symbols(params) do
      {:ok, {:lambda, names, body, env}}
    end
  end

  defp eval_expr({:list, [{:symbol, "if"}, cond, then_branch, else_branch]}, env) do
    with {:ok, {:number, cond_value}} <- eval_expr(cond, env) do
      if cond_value == 0 do
        eval_expr(else_branch, env)
      else
        eval_expr(then_branch, env)
      end
    end
  end

  defp eval_expr({:list, [head | tail]}, env) do
    with {:ok, fun} <- eval_expr(head, env),
         {:ok, args} <- eval_args(tail, env) do
      apply_fun(fun, args)
    end
  end

  defp eval_args(exprs, env) do
    exprs
    |> Enum.reduce_while({:ok, []}, fn expr, {:ok, acc} ->
      case eval_expr(expr, env) do
        {:ok, value} -> {:cont, {:ok, [value | acc]}}
        {:error, reason} -> {:halt, {:error, reason}}
      end
    end)
    |> case do
      {:ok, values} -> {:ok, Enum.reverse(values)}
      other -> other
    end
  end

  defp apply_fun({:builtin, fun}, args), do: fun.(args)

  defp apply_fun({:lambda, params, body, captured}, args) do
    if length(params) != length(args) do
      {:error, "引数の数が一致しません"}
    else
      new_bindings = Enum.zip(params, args) |> Map.new()
      merged = Map.merge(captured, new_bindings)
      eval_expr(body, merged)
    end
  end

  defp apply_fun({:number, _}, _), do: {:error, "数値は適用できません"}

  defp ensure_symbols(list) do
    list
    |> Enum.reduce_while({:ok, []}, fn
      {:symbol, name}, {:ok, acc} -> {:cont, {:ok, [name | acc]}}
      other, _ -> {:halt, {:error, "シンボルを期待しましたが #{inspect(other)} でした"}}
    end)
    |> case do
      {:ok, names} -> {:ok, Enum.reverse(names)}
      error -> error
    end
  end

  defp default_env do
    %{
      "+" => numeric2(&Kernel.+/2),
      "-" => numeric2(&Kernel.-/2),
      "*" => numeric2(&Kernel.*/2),
      "/" => numeric2(fn _a, 0 -> {:error, "0 で割れません"}; a, b -> {:ok, a / b} end),
      ">" => numeric2(fn a, b -> {:ok, if(a > b, do: 1.0, else: 0.0)} end),
      "=" => numeric2(fn a, b -> {:ok, if(a == b, do: 1.0, else: 0.0)} end)
    }
  end

  defp numeric2(fun) do
    {:builtin,
     fn
       [{:number, a}, {:number, b}] ->
         case fun.(a, b) do
           {:ok, result} -> {:ok, {:number, result}}
           {:error, reason} -> {:error, reason}
           result when is_number(result) -> {:ok, {:number, result}}
         end
       _ -> {:error, "数値 2 引数を期待します"}
     end}
  end
end

File: samples/language-impl-comparison/rust/mini_lisp.rs
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
enum Expr {
    Number(f64),
    Symbol(String),
    List(Vec<Expr>),
}

#[derive(Clone)]
enum Value {
    Number(f64),
    Lambda { params: Vec<String>, body: Expr, env: Env },
    Builtin(fn(&[Value]) -> Result<Value, String>),
}

type Env = HashMap<String, Value>;

pub fn eval(source: &str) -> Result<Value, String> {
    let tokens = tokenize(source);
    let (expr, rest) = parse_expr(&tokens, 0)?;
    if rest != tokens.len() {
        return Err("未消費トークンがあります".into());
    }
    let mut env = default_env();
    eval_expr(&expr, &mut env)
}

fn tokenize(source: &str) -> Vec<String> {
    source
        .replace('(', " ( ")
        .replace(')', " ) ")
        .split_whitespace()
        .map(|s| s.to_string())
        .collect()
}

fn parse_expr(tokens: &[String], index: usize) -> Result<(Expr, usize), String> {
    if index >= tokens.len() {
        return Err("入力が空です".into());
    }
    let token = &tokens[index];
    match token.as_str() {
        "(" => parse_list(tokens, index + 1),
        ")" => Err("対応しない閉じ括弧です".into()),
        t => Ok((parse_atom(t)?, index + 1)),
    }
}

fn parse_list(tokens: &[String], mut index: usize) -> Result<(Expr, usize), String> {
    let mut items = Vec::new();
    while index < tokens.len() {
        if tokens[index] == ")" {
            return Ok((Expr::List(items), index + 1));
        }
        let (expr, next) = parse_expr(tokens, index)?;
        items.push(expr);
        index = next;
    }
    Err("リストが閉じられていません".into())
}

fn parse_atom(token: &str) -> Result<Expr, String> {
    if let Ok(number) = token.parse::<f64>() {
        Ok(Expr::Number(number))
    } else {
        Ok(Expr::Symbol(token.to_string()))
    }
}

fn eval_expr(expr: &Expr, env: &mut Env) -> Result<Value, String> {
    match expr {
        Expr::Number(n) => Ok(Value::Number(*n)),
        Expr::Symbol(name) => env.get(name).cloned().ok_or_else(|| format!("未定義シンボル: {name}")),
        Expr::List(items) => {
            if items.is_empty() {
                return Err("空の式は評価できません".into());
            }
            let callee = eval_expr(&items[0], env)?;
            let mut args = Vec::new();
            for arg_expr in &items[1..] {
                args.push(eval_expr(arg_expr, env)?);
            }
            apply(callee, &args)
        }
    }
}

fn apply(callee: Value, args: &[Value]) -> Result<Value, String> {
    match callee {
        Value::Builtin(fun) => fun(args),
        Value::Lambda { params, body, mut env } => {
            if params.len() != args.len() {
                return Err("引数の数が一致しません".into());
            }
            for (param, arg) in params.into_iter().zip(args.iter().cloned()) {
                env.insert(param, arg);
            }
            eval_expr(&body, &mut env)
        }
        Value::Number(_) => Err("数値を関数適用できません".into()),
    }
}

fn default_env() -> Env {
    let mut env = HashMap::new();
    env.insert("+".into(), Value::Builtin(|args| numeric_op(args, |a, b| Ok(a + b))));
    env.insert("-".into(), Value::Builtin(|args| numeric_op(args, |a, b| Ok(a - b))));
    env.insert("*".into(), Value::Builtin(|args| numeric_op(args, |a, b| Ok(a * b))));
    env.insert("/".into(), Value::Builtin(|args| numeric_op(args, |a, b| {
        if b == 0.0 {
            Err("0 で割れません".into())
        } else {
            Ok(a / b)
        }
    })));
    env
}

fn numeric_op<F>(args: &[Value], op: F) -> Result<Value, String>
where
    F: Fn(f64, f64) -> Result<f64, String>,
{
    if args.len() != 2 {
        return Err("2 引数で呼び出してください".into());
    }
    match (&args[0], &args[1]) {
        (Value::Number(lhs), Value::Number(rhs)) => Ok(Value::Number(op(*lhs, *rhs)?)),
        _ => Err("数値以外を演算できません".into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn eval_add() {
        let result = eval("(+ 40 2)").unwrap();
        match result {
            Value::Number(n) => assert_eq!(n, 42.0),
            _ => panic!("数値が返る想定"),
        }
    }
}
